master/api_service.go

```diff
@@ -200,6 +200,94 @@ func (m *Server) setupAutoAllocation(w http.ResponseWriter, r *http.Request) {
 	sendOkReply(w, r, newSuccessHTTPReply(fmt.Sprintf("set DisableAutoAllocate to %v successfully", status)))
 }
 
+func (m *Server) forbidVolume(w http.ResponseWriter, r *http.Request) {
+	var (
+		status bool
+		name   string
+		err    error
+	)
+	metric := exporter.NewTPCnt(apiToMetricsName(proto.AdminVolForbidden))
+	defer func() {
+		doStatAndMetric(proto.AdminVolForbidden, metric, err, nil)
+		if err != nil {
+			log.LogErrorf("set volume forbidden failed, error: %v", err)
+		} else {
+			log.LogInfof("set volume forbidden to (%v) success", status)
+		}
+	}()
+	if name, err = parseAndExtractName(r); err != nil {
+		sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeParamError, Msg: err.Error()})
+		return
+	}
+	if status, err = parseAndExtractStatus(r); err != nil {
+		sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeParamError, Msg: err.Error()})
+		return
+	}
+	// step1: get vol, dp and mp
+	vol, err := m.cluster.getVol(name)
+	if err != nil {
+		sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeVolNotExists, Msg: err.Error()})
+		return
+	}
+	cmds := make(map[string]*RaftCmd)
+	// step2: update vol
+	oldForbiden := vol.Forbidden
+	vol.Forbidden = status
+	defer func() {
+		if err != nil {
+			vol.Forbidden = oldForbiden
+		}
+	}()
+	metadata, err := m.cluster.buildVolInfoRaftCmd(opSyncUpdateVol, vol)
+	if err != nil {
+		sendErrReply(w, r, newErrHTTPReply(err))
+		return
+	}
+	cmds[metadata.K] = metadata
+	dps := vol.cloneDataPartitionMap()
+	mps := vol.cloneMetaPartitionMap()
+	// step3: update dp
+	for _, dp := range dps {
+		old := dp.Forbidden
+		dp.Forbidden = status
+		metadata, err = m.cluster.buildDataPartitionRaftCmd(opSyncUpdateDataPartition, dp)
+		if err != nil {
+			sendErrReply(w, r, newErrHTTPReply(err))
+			return
+		}
+		cmds[metadata.K] = metadata
+		copyDp := dp
+		defer func() {
+			if err != nil {
+				copyDp.Forbidden = old
+			}
+		}()
+	}
+	// step4: update mp
+	for _, mp := range mps {
+		old := mp.Forbidden
+		mp.Forbidden = status
+		metadata, err = m.cluster.buildMetaPartitionRaftCmd(opSyncUpdateMetaPartition, mp)
+		if err != nil {
+			sendErrReply(w, r, newErrHTTPReply(err))
+			return
+		}
+		cmds[metadata.K] = metadata
+		copyMp := mp
+		defer func() {
+			if err != nil {
+				copyMp.Forbidden = old
+			}
+		}()
+	}
+	// step5: submit to raft
+	if err = m.cluster.syncBatchCommitCmd(cmds); err != nil {
+		sendErrReply(w, r, newErrHTTPReply(err))
+		return
+	}
+	sendOkReply(w, r, newSuccessHTTPReply(fmt.Sprintf("set volume forbidden to (%v) success", status)))
+}
+
```

Comment:
Based on the previous code, here's a function encapsulation。

Reply:
Ok